Launchpad: Experiment setup for Launchpad A/B test

daledupreez opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to https://github.com/Automattic/dotcom-forge/issues/8541

## Proposed Changes

* This exploratory PR looks at how complicated it might be to put together a PR that would allow us to create an A/B/C test around the fullscreen Launchpad. The idea for the test would be that we'd want to compare three behaviours:
   1. The current experience, where users see the fullscreen launchpad, and need to Skip it to get to Customer Home
   2. Instead of showing the fullscreen launchpad, we mark the launchpad as skipped, which should show it less obtrusively in Customer Home
   3. We always disable the fullscreen launchpad, so it's not displayed in the flow or Customer Home.

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

* To make the experiment possible: pbxNRc-3Yh-p2

## Testing Instructions

**Prerequisite**
- Run your sandbox and apply changes from https://github.com/Automattic/jetpack/pull/39237
- Proxy the {SITE_SLUG} you choose for testing
- Since we made changes on `apps/wpcom-block-editor/` we should follow:
  - Proxy `widgets.wp.com`
  - From the path `apps/wpcom-block-editor/` run the command `yarn dev --sync `
  - More info about syncing widget app with sandbox, find [here](https://github.com/Automattic/wp-calypso/blob/382b78199f323b152fbfa38a8c34af3e41d84166/apps/wpcom-block-editor/README.md). 
- In the active tab, open your developer tool and set the experiment variation as a session item. `window.sessionStorage.setItem( 'launchpad_removal_2024_experiment_variation', 'treatment' );`


### BUILD flow

**Scenario 1:**
* go to `/setup/site-setup?siteSlug={SITE_SLUG}`
* choose "Promote myself or business"
* select "Choose a theme"
* select a theme
* press "continue" button
* check if there is a home screen instead of the full-screen launchpad

**Scenario 2:**
- go to `/setup/site-setup?siteSlug={SITE_SLUG}`
- without goal selection, press "continue" button
- select "Design your own"
- select accordingly styles, pages, add additional content
- press WP icon button (top left corner)
- from the left sidebar navigation, click on the "Go to dashboard" back button
- check if there is a home screen instead of a full-screen launchpad

**Scenario 3:**
- make everything from scenario 2, but instead of WP icon
- make any change
- and press "Save"
- from the "Great progress" modal, click on the "Next Steps" button
- check if there is a home screen instead of a full-screen launchpad

---

### DESIGN FIRST flow

- go to `/setup/design-first/new-or-existing-site`
- select your site
- pick a design and press "continue"
- check if there is a home screen instead of a full-screen launchpad

---

### START WRITING flow

- go to `/setup/start-writing`
- select or create a site
- enter any content and "publish" the post
- from the "Great progress" modal, click on the "Next Steps" button
- check if there is a home screen instead of a full-screen launchpad

---

### NEWSLETTER flow
- go to `/setup/newsletter`
- select the site title, plan, and domain
- check if there is a home screen instead of a full-screen launchpad

---

### WRITE flow
- go to `/setup/site-setup?siteSlug={SITE_SLUG}`
- choose "Write & publish"
- select "Start writing" (Draft your first post)
- edit post content and "Publish"
- from the "Great progress" modal, click on the "Next Steps" button
- check if there is a home screen instead of a full-screen launchpad
<img width="1006" alt="Screenshot 2024-09-06 at 17 43 23" src="https://github.com/user-attachments/assets/b659be18-b3bb-4332-bc9a-a28ceef53d8e">

---

### FREE flow
- go to `/setup/free`
- enter a site name
- select theme and styles
- check if there is a home screen instead of a full-screen launchpad

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

daledupreez committed:
Explore experiment setup for Launchpad A/B/C test

daledupreez committed:
Update experiment name

daledupreez committed:
Add siteId argument

daledupreez committed:
Switch back to treatment|control and use clearer variables

bogiii committed:
Integrate launchpad decider into build flow

bogiii committed:
Integrate launchpad decider into update-design (design-first) flow

bogiii committed:
Use flow name from const

bogiii committed:
Integrate launchpad decider into start-writing flow

bogiii committed:
Integrate launchpad decider into newsletter flow

bogiii committed:
Create useExitFlow hook to reduce code duplication

bogiii committed:
Use exitFlow from the hook

bogiii committed:
Extend useExitFlow hook

bogiii committed:
Integrate useExitFlow hook

bogiii committed:
Integrate launchpad decider into design-first flow

bogiii committed:
Fix bug discovered by unit test

bogiii committed:
Fix code-style

bogiii committed:
Apply experiment into controller redirection logic

bogiii committed:
Introduce session storage item supporting experiment testing

bogiii committed:
Remove unnecessary init launchpad state

ouikhuan reviewed (approved):
All the flow listed works as describe, great work!

daledupreez commented on client/my-sites/customer-home/controller.jsx:
```diff
@@ -67,7 +68,17 @@ export async function maybeRedirect( context, next ) {
+		const shouldShowLaunchpad =
+			'treatment' !== experimentAssignment?.variationName &&
+			// for testing/development purposes we can use sessionStorage to force the treatment
+			'treatment' !== window.sessionStorage.getItem( 'launchpad_experiment_variation' );
```
It may be worth using a more specific name to avoid accidental re-use.
```suggestion
			'treatment' !== window.sessionStorage.getItem( 'launchpad_removal_2024_experiment_variation' );
```

bogiii replied:
I agree with you. I'll update it here and in [the calypso PR](https://github.com/Automattic/jetpack/pull/39237) as well.

daledupreez commented on client/landing/stepper/declarative-flow/site-setup-flow.ts:
```diff
@@ -161,6 +166,45 @@ const siteSetupFlow: Flow = {
+				return 'full';
+			}
+
+			if ( launchpadExperimentAssigment.variationName === 'treatment_no_launchpad' ) {
+				return 'off';
+			}
+
+			if ( launchpadExperimentAssigment.variationName === 'treatment_skip_launchpad' ) {
+				return 'skipped';
+			}
```
I think this should be looking at `'treatment'` for the variation, and removing this old logic for the A/B/C test.
```suggestion
			if ( launchpadExperimentAssigment.variationName === 'treatment' ) {
				return 'skipped';
			}
```

daledupreez replied:
Also, it feels like this logic should also be exposed/shared with `useLaunchpadDecider()`.

daledupreez commented on client/landing/stepper/declarative-flow/internals/hooks/use-launchpad-decider/index.tsx:
```diff
@@ -0,0 +1,72 @@
+		! isLoadingExperiment &&
+		( 'treatment' === experimentAssignment?.variationName ||
+			// for testing/development purposes we can use sessionStorage to force the treatment
+			'treatment' === window.sessionStorage.getItem( 'launchpad_experiment_variation' ) );
```
It may be worth wrapping this access in a function so we can make it a no-op and/or change the key name easily.

bogiii replied:
Absolutely agree. Done!

daledupreez commented on apps/wpcom-block-editor/src/wpcom/features/redirect-onboarding-user-after-publishing-post.js:
```diff
@@ -25,6 +25,7 @@ export function RedirectOnboardingUserAfterPublishingPost() {
 		hasStartWritingFlowQueryArg &&
 		'full' === launchpadScreen &&
 		currentPostType === 'post';
+	const postFlowUrl = getQueryArg( window.location.search, 'postFlowUrl' );
```
Non-block nit: we could also call this `postPublishUrl`, but I don't think it's worth much rework.

daledupreez reviewed (commented):
Overall, this is looking good to me. The only exception is the code block looking at the older variation names, which should be quick to fix. Beyond that, I have some minor, non-blocking suggestions and thoughts. Also, I can't approve as I was the original author.

bogiii committed:
Rename the session item name to match the experiment

bogiii committed:
Remove obsolete A/B/C logic

bogiii committed:
Move logic for getting launchpad state to the separate function

bogiii committed:
Refactor: Move customer home logic to a separate function

- Extracted logic to determine if customer home should be shown into a new function `shouldShowCustomerHome`.
- Updated `useLaunchpadDecider` to use the new `shouldShowCustomerHome` function.

necronet commented on client/landing/stepper/declarative-flow/internals/hooks/use-launchpad-decider/index.tsx:
```diff
@@ -0,0 +1,114 @@
+		! isLoadingExperiment &&
+		( 'treatment' === experimentAssignment?.variationName ||
+			// for testing/development purposes we can use sessionStorage to force the treatment
+			'treatment' ===
+				window.sessionStorage.getItem( 'launchpad_removal_2024_experiment_variation' ) )
```
will this eventually get removed from the code changes I see we can force the experiment variation through client side ?

bogiii replied:
The session item is only for our dev/testing purposes. We are setting it manually via the developer tool. I didn't find a better way to test it.
Once we start the experiment, I can remove it in an additional PR.

bogiii replied:
Since the experiment is running, I created PR for removing dev/testing piece of code (we don't need it anymore): #94764

necronet replied:
Thanks for the update!

necronet commented on client/landing/stepper/declarative-flow/internals/hooks/use-launchpad-decider/index.tsx:
```diff
@@ -0,0 +1,114 @@
+			}
+
+			navigate( `launchpad?siteSlug=${ siteSlug }&siteId=${ siteId }` );
+			return;
```
do we need this return ?

bogiii replied:
True, we don't need it since it's the last item in the function body.

necronet commented on client/landing/stepper/declarative-flow/internals/hooks/use-launchpad-decider/index.tsx:
```diff
@@ -0,0 +1,114 @@
+): boolean {
+	return (
+		! isLoadingExperiment &&
+		( 'treatment' === experimentAssignment?.variationName ||
```
I'm a bit confused because I though this was meant to be a:
>  A/B/C test around the fullscreen Launchpad.

Yet I see only `treatment` been used throughout the changes

bogiii replied:
The experiment is an A/B test, so we have **control** and **treatment** variants.

Control - the onboarding flow remains the same as it is
Treatment - skips the fullscreen onboarding launchpad step

bogiii committed:
Remove unnecessary returns
